File: Data Structures & Algorithms/decode-ways/submission-2.py
class Solution:
    def numDecodings(self, s: str) -> int:
        n = len(s)
        dp = {n : 1}
        res = 0

        def dfs(i):
            if i in dp:
                return dp[i]
            if s[i] == '0':
                return 0
            
            res = dfs(i + 1)
            if (i < n - 1 and (s[i] == '1' or s[i] == '2' and int(s[i + 1]) < 7)):
                res += dfs(i + 2)
            
            dp[i] = res
            return res
        
        return dfs(0)

File: Data Structures & Algorithms/prim/submission-2.java
class Solution {
    public int minimumSpanningTree(List<List<Integer>> edges, int n) {
        Map<Integer, List<int[]>> adj = new HashMap<>();
        Set<Integer> visit = new HashSet<>();
        Queue<int[]> q = new PriorityQueue<>((a, b) -> Integer.compare(a[0], b[0]));
        for (List<Integer> edge : edges)
        {
            int u = edge.get(0);
            int v = edge.get(1);
            int w = edge.get(2);

            adj.computeIfAbsent(u, k -> new ArrayList<int[]>()).add(new int[]{w, v});
            adj.computeIfAbsent(v, k -> new ArrayList<int[]>()).add(new int[]{w, u});
        }

        q.offer(new int[]{0, 0});
        int res = 0;

        while (!q.isEmpty() && visit.size() < n)
        {
            int[] data = q.poll();
            int w = data[0], node = data[1];

            if (visit.contains(node))
            {
                continue;
            }

            visit.add(node);
            res += w;

            for (int[] neighbor : adj.getOrDefault(node, new ArrayList<>()))
            {
                int new_weight = neighbor[0];
                int nei = neighbor[1];
                if (!visit.contains(nei))
                {
                    q.offer(new int[]{new_weight, nei});
                }
            }
        }

        return visit.size() < n ? -1 : res;
    }
}    
